File: src/boj/boj_10844.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class boj_10844 {
    static final int MOD = 1_000_000_000;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][] dp = new int[n + 1][10];
        for (int i = 1; i < 10; i++) {
            dp[1][i] = 1;
        }
        for (int i = 2; i <= n; i++) {
            for (int j = 0; j < 10; j++) {
                if (0 <= j - 1) {
                    dp[i][j] += dp[i - 1][j - 1];
                    dp[i][j] %= MOD;
                }
                if (j + 1 < 10) {
                    dp[i][j] += dp[i - 1][j + 1];
                    dp[i][j] %= MOD;
                }
            }
        }

        int answer = 0;
        for (int i = 0; i < 10; i++) {
            answer += dp[n][i];
            answer %= MOD;
        }
        System.out.println(answer);
        br.close();
    }
}

File: src/boj/boj_10844.kt
package boj

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val mod = 1000000000
    val n = readLine().toInt()
    val cache = Array(n + 1) { IntArray(10) }
    for (i in 0..9) {
        cache[1][i] = if (i == 0) 0 else 1
    }

    for (i in 2..n) {
        for (j in 0..9) {
            cache[i][j] = when (j) {
                0 -> cache[i - 1][1]
                9 -> cache[i - 1][8]
                else -> (cache[i - 1][j - 1] + cache[i - 1][j + 1]) % mod
            }
        }
    }

    var answer: Long = 0
    for (i in 0..9) {
        answer += cache[n][i]
        answer %= mod
    }
    println(answer)

    close()
}
